handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -0,0 +1,1951 @@
+                            case SSL.SSL_ERROR_WANT_WRITE:
+                                // SSL_ERROR_WANT_WRITE typically means that the underlying transport is not writable
+                                // and we should set the "want write" flag on the selector and try again when the
+                                // underlying transport is writable [1]. However we are not directly writing to the
+                                // underlying transport and instead writing to a BIO buffer. The OpenSsl documentation
+                                // says we should do the following [1]:
+                                //
+                                // "When using a buffering BIO, like a BIO pair, data must be written into or retrieved
+                                // out of the BIO before being able to continue."
+                                //
+                                // So we attempt to drain the BIO buffer below, but if there is no data this condition
+                                // is undefined and we assume their is a fatal error with the openssl engine and close.
+                                // [1] https://www.openssl.org/docs/manmaster/ssl/SSL_write.html
+                                pendingNetResult = readPendingBytesFromBIO(dst, bytesConsumed, bytesProduced, status);
+                                return pendingNetResult != null ? pendingNetResult : NEED_WRAP_CLOSED;
+                            default:
+                                // Everything else is considered as error
+                                throw shutdownWithError("SSL_write");
+                        }
+                    }
+                }
+            }
+            // We need to check if pendingWrittenBytesInBIO was checked yet, as we may not checked if the srcs was
+            // empty, or only contained empty buffers.
+            if (bytesConsumed == 0) {
+                SSLEngineResult pendingNetResult = readPendingBytesFromBIO(dst, 0, bytesProduced, status);
+                if (pendingNetResult != null) {
+                    return pendingNetResult;
+                }
+            }
+
+            return newResult(bytesConsumed, bytesProduced, status);
+        }
+    }
+
+    /**
+     * Log the error, shutdown the engine and throw an exception.
+     */
+    private SSLException shutdownWithError(String operations) {
+        String err = SSL.getLastError();
+        return shutdownWithError(operations, err);
+    }
+
+    private SSLException shutdownWithError(String operation, String err) {
+        if (logger.isDebugEnabled()) {
+            logger.debug("{} failed: OpenSSL error: {}", operation, err);
+        }
+
+        // There was an internal error -- shutdown
+        shutdown();
+        if (handshakeState == HandshakeState.FINISHED) {
+            return new SSLException(err);
+        }
+        return new SSLHandshakeException(err);
+    }
+
+    public final SSLEngineResult unwrap(
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "unwrap" is 55 which is greater than 25 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)